### PR TITLE
Nascent support for subdomains

### PR DESCRIPTION
**Pull request by miklos1**
##### Limitations:
 * Cell integrals only (`dx`)
 * Whole cells only (not cutting of cells at this stage)
 * For an extruded mesh:
   * Columns are monolithic (either all or none of the cells in a column are in the subdomain)
   * API not stable yet

##### Features:
 * Subdomain defined as geometric expression (UFL-based)
 * Works on 0-forms, 1-forms and 2-forms

##### Usage:
    ...
    x = mesh.coordinates
    sd = SubDomainData(x[0] < 0.5)
    assemble(f*dx(subdomain_data=sd))

### CODE REVIEW

**Comment by miklos1** [comment]
I need to add some parallel tests as well.

**Comment by miklos1** [comment]
OK, I think I fixed it for parallel and added parallel tests.

File: firedrake/mesh.py [hunk 1]
```diff
@@ -1151,3 +1151,19 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
                                     layer_height, extrusion_type="radial", kernel=kernel)
 
     return self
+
+
+def SubDomainData(geometric_expr):
```
Comment (wence-) [inline firedrake/mesh.py]: Docstring?

File: firedrake/assemble.py [hunk 2]
```diff
@@ -231,8 +231,8 @@ def vec(testmap, i):
     # assemble it.
     def thunk(bcs):
         zero_tensor()
-        for (i, j), integral_type, subdomain_id, coords, coefficients, needs_orientations, kernel in kernels:
-            m = coords.function_space().mesh()
+        for (i, j), integral_type, m, subdomain_data, subdomain_id, coefficients, needs_orientations, kernel in kernels:
+            coords = m.coordinates
             if needs_orientations:
```
Comment (wence-) [inline firedrake/assemble.py]: Can you raise a NotImplementedError here if subdomain_data is provided and the integral_type is not `cell`?

**Comment by dham** [comment]
How does this compare with the syntax for similar functionality in DOLFIN?

**Comment by wence-** [comment]
It would be nice to have some documentation on the usage of this feature.

**Comment by wence-** [comment]
The dolfin syntax is shown in http://fenicsproject.org/documentation/dolfin/1.6.0/python/demo/documented/subdomains-poisson/python/documentation.html.  So this is not that close yet.  OTOH, these aren't yet fully-fledged subdomains (and the "class" name is different) so maybe it's not needed to be addressed yet.

**Comment by miklos1** [comment]
@dham: The two syntaxes don't overlap.

What we want to achieve in the long term is far more general than what DOLFIN currently has, and in the light of this I think that trying to adopt and extend DOLFIN's syntax isn't a good idea.

I admit that supporting DOLFIN's syntax would be beneficial for compatibility reasons. However, that wouldn't be very efficient (iteration over mesh in Python), and I don't think it's very urgent yet.

DOLFIN's syntax seems to be based on the premises that:
 * Coordinates are special, not just a `Function`
 * We have a C++ library wrapped in Python

**Comment by wence-** [comment]
Having thought more, I'm now opposed to the proposed UFL change.  I have pushed a change adding the subdomain data to the assembly cache key.  This means that the proposed UFL changes are no longer necessary.

File: firedrake/mesh.py (hunk 1 above)
Comment (miklos1) [inline firedrake/mesh.py]: Added.

File: firedrake/mesh.py [hunk 3]
```diff
@@ -1151,3 +1151,29 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
...
+    """Creates a subdomain data object from a geometric UFL expression.
```
Comment (wence-) [inline firedrake/mesh.py]: I don't know what a geometric UFL expression is.

**Comment by miklos1** [comment]
Can we merge this now?

**Pull request merged**